<!--
  ~ Copyright 2015-2016 OpenCB
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~     http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->

<link rel="import" href="opencga-variant-filter.html">
<link rel="import" href="opencga-variant-interpretation-grid.html">
<link rel="import" href="opencga-variant-interpretation-detail.html">
<link rel="import" href="../opencga-active-filters.html">

<dom-module id="opencga-variant-family-analysis">
    <template>
        <style>
            .qtip-family-class {
                max-width: 480px;
                font-size: 12px;
            }
        </style>

        <div class="row">
            <div class="col-md-12">
                <h4 id="{{prefix}}Title" class="form-section-title" style="margin: 5px 0px;padding: 0px 10px">
                    {{_config.title}}
                    <span class='family-info-icon' style="color: #337ab7; padding-left: 20px">
                        <i class='fa fa-info-circle' aria-hidden='true'></i>
                    </span>
                </h4>
            </div>

            <template is="dom-if" if="{{!existParent}}">
                <div class="col-md-8 col-md-offset-2" style="padding: 20px">
                    <div class="alert alert-warning" role="alert" id="{{prefix}}Warning" style="font-size: 14px;margin-bottom: 10px">
                        <span style="font-style: italic; font-weight: bold">Compound Heterozygous Analysis </span> require at least one parent,
                        no parents found for proband <span style="font-weight: bold; margin-left: 5px">{{clinicalAnalysis.proband.id}}</span>.
                    </div>
                </div>
            </template>

            <template is="dom-if" if="{{existParent}}">
                <div class="col-md-12" style="padding: 10px 0px">
                    <div class="col-md-2">
                        <opencga-variant-filter opencga-session="{{opencgaSession}}"
                                                query="{{query}}"
                                                clinical-analysis="{{clinicalAnalysis}}"
                                                cellbase-client="{{cellbaseClient}}"
                                                population-frequencies="{{populationFrequencies}}"
                                                consequence-types="{{consequenceTypes}}"
                                                on-querychange="onVariantFilterChange"
                                                on-querysearch="onVariantFilterSearch" on-samplechange="onSampleChange"
                                                on-inheritancemode="_onInheritanceMode"
                                                config="{{_config.filter}}">
                        </opencga-variant-filter>
                    </div>

                    <div class="col-md-10">
                        <opencga-active-filters opencga-client="{{opencgaSession.opencgaClient}}"
                                                default-study="{{opencgaSession.study.alias}}"
                                                query="{{preparedQuery}}"
                                                refresh="{{executedQuery}}"
                                                filters="{{_config.filter.examples}}"
                                                on-filterchange="onActiveFilterChange"
                                                on-clear="onActiveFilterClear"
                                                filter-bioformat="VARIANT" alias="{{_config.activeFilterAlias}}"
                                                genotype-samples="{{genotypeSamples}}" mode-inheritance="[[modeInheritance]]"
                                                config="{{_config.activeFilters}}">
                        </opencga-active-filters>

                        <div style="padding-top: 5px">
                            <opencga-variant-interpretation-grid opencga-session="{{opencgaSession}}"
                                                                 clinical-analysis="{{clinicalAnalysis}}"
                                                                 query="{{executedQuery}}"
                                                                 consequence-types="{{consequenceTypes}}"
                                                                 population-frequencies="{{populationFrequencies}}"
                                                                 protein-substitution-scores="{{proteinSubstitutionScores}}"
                                                                 on-selected="selectedGene"
                                                                 on-selectvariant="onSelectVariant"
                                                                 on-checkvariant="onCheckVariant"
                                                                 on-setgenomebrowserposition="onGenomeBrowserPositionChange"
                                                                 config="{{config.grid}}">
                            </opencga-variant-interpretation-grid>

                            <!-- Bottom tabs with detailed variant information -->
                            <opencga-variant-interpretation-detail opencga-session={{opencgaSession}}
                                                                   cellbase-client={{cellbaseClient}}
                                                                   variant={{variant}}
                                                                   clinical-analysis={{clinicalAnalysis}}
                                                                   consequence-types={{consequenceTypes}}
                                                                   protein-substitution-scores="{{proteinSubstitutionScores}}"
                                                                   config={{config.detail}}>
                            </opencga-variant-interpretation-detail>
                        </div>
                    </div>
                </div>
            </template>
        </div>
    </template>

    <script>
        class OpencgaVariantFamilyAnalysis extends Polymer.Element {

            static get is() {
                return "opencga-variant-family-analysis"
            };

            static get properties() {
                return {
                    opencgaSession: {
                        type: Object
                    },
                    clinicalAnalysis: {
                        type: Object,
                        // observer: "clinicalAnalysisObserver"
                    },
                    query: {
                        type: Object,
                        observer: "queryObserver"
                    },
                    mode: {
                        type: String,
                        value: "interactive"
                    },
                    cellbaseClient: {
                        type: Object
                    },
                    active: {
                        type: Boolean,
                        value: true,
                        observer: "activeObserver"
                    },
                    consequenceTypes: {
                        type: Object
                    },
                    populationFrequencies: {
                        type: Object
                    },
                    proteinSubstitutionScores: {
                        type: Object
                    },
                    config: {
                        type: Object
                    }
                }
            }

            _attachDom(dom) {
                this.appendChild(dom);
            }

            connectedCallback() {
                super.connectedCallback();

                // this._addTooltip();
            }

            constructor() {
                super();

                // Set status and init private properties
                this._init();
            }

            /**
             * Init the variables you need, keep in mind this is executed before the actual Polymer properties exist
             */
            _init() {
                // All id fields in the template must start with prefix, this allows components to be instantiated more than once
                this.prefix = "ovdna" + Utils.randomString(6);

                this.existParent = true;
                this._samples = [];
                this._sampleIds = [];
                this._individualToSampleIds = {};

                // Initially we set the default config, this will be overridden if 'config' is passed
                // this._config = this.getDefaultConfig();
                // debugger
            }

            static get observers() {
                return [
                    "propertyObserver(opencgaSession, clinicalAnalysis)"
                ];
            }

            propertyObserver(opencgaSession, clinicalAnalysis) {
                let defaultConfig = this.getDefaultConfig();
                let _config = Object.assign({}, defaultConfig, JSON.parse(JSON.stringify(this.config)));
                _config.title = defaultConfig.title;
                _config.tooltip = defaultConfig.tooltip;
                if (this.mode !== "interactive") {
                    _config.filter.menu.skipSubsections.push("sample");
                }
                this._config = _config;
                this._addTooltip();

                if (UtilsNew.isNotUndefinedOrNull(this.opencgaSession) && UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis)) {
                    this._calculateSamples();

                    if (this.mode !== "interactive" && UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis.roleToProband)) {
                        let _existParent = false;
                        for (let individualId of Object.keys(this.clinicalAnalysis.roleToProband)) {
                            if (this.clinicalAnalysis.roleToProband[individualId] === "FATHER"
                                || this.clinicalAnalysis.roleToProband[individualId] === "MOTHER") {
                                if (UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis.files)
                                    && UtilsNew.isNotEmptyArray(this.clinicalAnalysis.files[this._individualToSampleIds[individualId]])) {
                                    for (let file of this.clinicalAnalysis.files[this._individualToSampleIds[individualId]]) {
                                        if (file.format === "VCF" && UtilsNew.isNotUndefinedOrNull(file.index.status)
                                            && file.index.status.name === "READY") {
                                            _existParent = true;
                                            break;
                                        }
                                    }
                                }
                            }
                        }
                        this.existParent = _existParent;
                    }

                    // Init query objects every time there is a new session or clinical analysis
                    let _query = Object.assign({}, this.query, {
                        study: this.opencgaSession.study.fqn,
                        includeSample: this._sampleIds.join(",")
                    });
                    switch (this.mode) {
                        case "interactive":
                            let _genotypes = [];
                            for (let sampleId of this._sampleIds) {
                                if (this.clinicalAnalysis.proband.samples[0].id === sampleId) {
                                    _genotypes.push(sampleId + ":0/1,1/1");
                                } else {
                                    _genotypes.push(sampleId + ":0/0,0/1,1/1");
                                }
                            }
                            _query.genotype = _genotypes.join(";");
                            break;
                        case "compoundHeterozygous":
                            _query.biotype = "protein_coding";
                            _query.ct = "missense_variant," + this._config.filter.lof.join(",");
                            break;
                    }

                    // Reset query when new session or clinical analysis is provided
                    this.query = _query;
                }
            }

            queryObserver() {
                if (UtilsNew.isNotUndefinedOrNull(this.query)) {
                    this.preparedQuery = this._prepareQuery(this.query);
                    this.executedQuery = this.preparedQuery;
                }
            }

            activeObserver() {
                if (this.active) {
                    // this.executedQuery = Object.assign({}, this.executedQuery);
                }
            }


            onVariantFilterChange(e) {
                this.preparedQuery = this._prepareQuery(e.detail.query);
            }

            onVariantFilterSearch(e) {
                this.preparedQuery = this._prepareQuery(e.detail.query);
                this.executedQuery = this.preparedQuery;
            }

            onActiveFilterChange(e) {
                this.query = e.detail;
            }

            onActiveFilterClear() {
                this.query = Object.assign({}, {
                    study: this.opencgaSession.study.fqn
                });
            }

            onSelectVariant(e) {
                this.variant = e.detail.variant;
            }

            _prepareQuery(query) {
                let _query = Object.assign({}, query, {
                    study: this.opencgaSession.study.fqn,
                    includeSample: this._sampleIds.join(",")
                });

                if (this.mode === "interactive") {
                    if (UtilsNew.isNotUndefinedOrNull(_query.genotype)) {
                        delete _query.sample;
                    } else {
                        if (UtilsNew.isUndefinedOrNull(_query.sample)) {
                            _query.sample = this._sampleIds.join(",");
                        }
                    }
                    delete _query.family;
                    return _query;
                } else {
                    if (UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis) && UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis.family)) {
                        _query = Object.assign({}, _query, {
                            // Replace family and familyDisorder with clinicalAnalysisId
                            family: this.clinicalAnalysis.family.id,
                            familyDisorder: this.clinicalAnalysis.disorder.id,
                            familySegregation: this.mode,
                        });
                        delete _query.genotype;
                        delete _query.sample;

                        // if (this.mode === "compoundHeterozygous") {
                        //     _query.biotype = "protein_coding";
                        //     _query.ct = "missense_variant," + this._config.filter.lof.join(",");
                        // }

                        return _query;
                    }
                }
            }

            _calculateSamples() {
                let _samples = [];
                let _sampleIds = [];
                let _individualToSampleIds = {};
                if (UtilsNew.isNotUndefinedOrNull(this.clinicalAnalysis.family)
                    && UtilsNew.isNotEmptyArray(this.clinicalAnalysis.family.members)) {
                    this.clinicalAnalysis.family.members.forEach(individual => {
                        if (UtilsNew.isNotEmptyArray(individual.samples)) {
                            _samples.push(individual.samples[0]);
                            _sampleIds.push(individual.samples[0].id);
                            _individualToSampleIds[individual.id] = individual.samples[0].id;
                        }
                    });
                }
                this._samples = _samples;
                this._sampleIds = _sampleIds;
                this._individualToSampleIds = _individualToSampleIds;
            }

            _addTooltip() {
                $(".family-info-icon").qtip({
                    content: {
                        title: "information",
                        text: this._config.tooltip,
                    },
                    position: {
                        target: "mouse",
                        // my: (config !== undefined && config.position !== undefined && config.position.my !== undefined) ? config.position.my : "top left",
                        adjust: {
                            x: 2, y: 2,
                            mouse: false
                        }
                    },
                    style: {
                        classes: "qtip-light qtip-rounded qtip-shadow qtip-family-class",
                        width: "480px",
                    },
                    show: {
                        delay: 250
                    },
                    hide: {
                        fixed: true,
                        delay: 300
                    }
                });
            }

            getDefaultConfig() {
                switch (this.mode) {
                    case "interactive":
                        return {
                            title: "Interactive Variant Interpretation Analysis",
                            tooltip: `<p>
                                        <span style="font-weight: bold">Interactive Variant</span> analysis allows clinicians to filter
                                        variants from the left menu and perform a clinical analysis. Results can be stored as an Interpretation.
                                      </p>`,
                            activeFilters: {
                                alias: {
                                    // Example:
                                    // "region": "Region",
                                },
                                complexFields: ["genotype"],
                                hiddenFields: ["study", "sample", "includeSample"]
                            },
                            genomeBrowser: {
                                showTitle: false
                            }
                        };
                    case "compoundHeterozygous":
                        return {
                            title: "Compound Heterozygous Variant Analysis",
                            tooltip: `<p>
                                        <span style="font-weight: bold">Compound Heterozygous Variant</span> in medical genetics is the
                                        condition of having two heterozygous recessive alleles at a particular gene locus that can cause a
                                        genetic disease in a heterozygous state, one on each chromosome of a pair. You can find more
                                        information at <a href="https://en.wikipedia.org/wiki/Compound_heterozygosity" target="_blank">Wikipedia</a>
                                      </p>
                                      <p>
                                        Compound Heterozygous variant analysis sets automatically the <span style="font-weight: bold">genotype</span>
                                        and <span style="font-weight: bold">biotype (protein_coding)</span> filters. However,
                                        you can use any other selected filter from the left menu. Note that finding Compound Heterozygous variants
                                        can take few seconds depending on the OpenCGA server installation.
                                      </p>`,
                            activeFilters: {
                                alias: {
                                    // "region": "Region",
                                },
                                complexFields: ["genotype"],
                                hiddenFields: ["study", "family", "familyDisorder", "familySegregation", "includeSample"],
                                lockedFields: [{id: "biotype", message: "'biotype' filter is mandatory in Compound Heterozygous analysis"}]
                            }
                        };
                    case "deNovo":
                        return {
                            title: "de Novo Variant Analysis",
                            tooltip: `<p>
                                        <span style="font-weight: bold">de Novo Variant</span> is a genetic alteration that is present
                                        for the first time in one family member as a result of a variant (or mutation) in a germ cell (egg or sperm)
                                        of one of the parents, or a variant that arises in the fertilized egg itself during early embryogenesis.
                                    </p>
                                    <p>
                                        de Novo variant analysis sets automatically the <span style="font-weight: bold">genotype</span> filter.
                                        However, you can use any other selected filter from the left menu. Note that finding de Novo variants can take few seconds
                                        depending on the OpenCGA server installation.
                                    </p>`,
                            activeFilters: {
                                alias: {
                                    // "region": "Region",
                                },
                                complexFields: ["genotype"],
                                hiddenFields: ["study", "family", "familyDisorder", "familySegregation", "includeSample"]
                            }
                        };
                }
            }
        }

        customElements.define(OpencgaVariantFamilyAnalysis.is, OpencgaVariantFamilyAnalysis);
    </script>
</dom-module>